src/stress/adapter/device_allocation.spec.ts

```diff
@@ -0,0 +1,264 @@
+              storeOp: 'store',
+            },
+          ],
+        });
+        pass.setPipeline(pipeline);
+        pass.setBindGroup(0, bindgroup);
+        pass.draw(kSize * kSize);
+        pass.endPass();
+        commands.push(encoder.finish());
+      }
+    }
+    return { device, objects: commands };
+  }
+
+  async createDeviceAndBuffers(adapter: GPUAdapter) {
+    // Currently we just allocate 2G of memory using 512MB blocks. We may be able to
+    // increase this to hit OOM instead, but on integrated GPUs on Metal, this can cause
+    // kernel panics at the moment, and it can greatly increase the time needed.
+    const kTotalMemorySize = 2 * 1024 * 1024 * 1024;
+    const kMemoryBlockSize = 512 * 1024 * 1024;
+    const kMemoryBlockData = new Uint8Array(kMemoryBlockSize);
+
+    const device: GPUDevice = await adapter.requestDevice();
+    const buffers = [];
+    for (let memory = 0; memory < kTotalMemorySize; memory += kMemoryBlockSize) {
+      const buffer = device.createBuffer({
+        size: kMemoryBlockSize,
+        usage: GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_DST,
+      });
+
+      // Write out to the buffer to make sure that it has backing memory.
+      device.queue.writeBuffer(buffer, 0, kMemoryBlockData, 0, kMemoryBlockData.length);
+      buffers.push(buffer);
+    }
+    return { device, objects: buffers };
+  }
+}
+
+export const g = makeTestGroup(DeviceAllocationTests);
+
+g.test('coexisting')
+  .desc(`Tests allocation of many coexisting GPUDevice objects.`)
+  .params(u => u.combine('adapterType', kAdapterTypes))
+  .fn(async t => {
+    const { adapterType } = t.params;
+    const adapter = await getGPU().requestAdapter(kAdapterTypeOptions[adapterType]);
+    assert(adapter !== null, 'Failed to get adapter.');
+
+    // Based on Vulkan comformance test requirement to be able to create multiple devices.
+    const kNumDevices = 5;
+
+    const devices = [];
+    for (let i = 0; i < kNumDevices; ++i) {
+      const device: GPUDevice = await adapter.requestDevice();
+      devices.push(device);
+    }
+  });
+
+g.test('continuous,with_destroy')
+  .desc(
+    `Tests allocation and destruction of many GPUDevice objects over time. Device objects
+are sequentially requested with a series of device allocated objects created on each
+device. The devices are then destroyed to verify that the device and the device allocated
+objects are recycled over a very large number of iterations.`
+  )
+  .params(u => u.combine('adapterType', kAdapterTypes))
+  .fn(async t => {
+    const { adapterType } = t.params;
+    const adapter = await getGPU().requestAdapter(kAdapterTypeOptions[adapterType]);
+    assert(adapter !== null, 'Failed to get adapter.');
+
+    // Since devices are being destroyed, we should be able to create many devices.
+    const kNumDevices = 100;
+    const kFunctions = [
+      t.createDeviceAndBuffers,
+      t.createDeviceAndComputeCommands,
+      t.createDeviceAndRenderCommands,
+    ];
```

Comment:
nit: probably best to put these each as a param, or a subcase param. Overall, I think this test and the following test can have params deviceOnly, deviceAndBuffers, deviceAndComputeCommands, deviceAndRenderCommands

Reply:
Hmm not entirely sure what you mean, but I am hesitant to make them params because the test itself takes on the order of 3-5 minutes already, so having a combination of things could make the suite take a long time. At the same time though, I don't want to have to decrease the number of iterations per run since the large number of iterations is important for testing OOM situations.

I might leave them as is for now?